### PR TITLE
v2.0.0 Beta - Updated to Jest 27, added deeper implementations for both req and res, refactored and further improved test coverage.

### LINKED ISSUE #58: .status() not chaining
I noticed that when my app had a chained response like so:
```javascript
response.status(204).send();
```
the tests would fail because `status()` is returning undefined in the tests, while really it returns a reference to `this` enabling it to chain.

To get it to chain properly in my tests I had to implement the following:
```javascript
// Top of test file
const mockedStatusFunc = jest.fn();
const mockedResponse = getMockRes({ status: mockedStatusFunc });
// In my tests
mockedStatusFunc.mockReturnThis();
```
I did notice though that it worked just fine if I didn't chain the functions together which is what led me to this conclusion.
```javascript
// Worked
response.status(204);
response.send();
// Doesn't work
response.status(204).send();
```

### LINKED ISSUE #72: Implement Readable mocks on Request
Mock implementations for the Express `Request` have not been implemented deep enough. These need to be implemented to prevent type errors when used.

Express Request hierarchy:
- `express.Request` ✔
- `http.IncomingMessage` ✔
- `stream.Readable` ❌
- `event.EventEmitter` ❌

Implement the properties and functions for `stream.Readable` and `event.EventEmitter`.

### LINKED ISSUE #73: Implement Writable mocks on Response
Mock implementations for the Express `Response` have not been implemented deep enough. These need to be implemented to prevent type errors when used.

Express Response hierarchy:
- `express.Response` ✔
- `http.ServerResponse` ✔
- `http.OutgoingMessage` ✔
- `stream.Writable` ❌
- `event.EventEmitter` ❌

Implement the properties and functions for `stream.Writable` and `event.EventEmitter`.

### PR DESCRIPTION
**Pull request by aboyce**
- Updated to `Jest` v27
- Updated dependencies
- `tmpl` from `1.0.4` to `1.0.5` (@dependabot)
- Added in tests for the return type of the `res` functions (to see if they are chainable or not)
- Updated CI to use new NodeJS versions
- Updated `snyk` to use new library, removed old
- Improve types for `request` and `response` (use library types rather than copying)
- Test structure refactor
    - All `request` properties have a test for the default value
    - All `request` properties have a test for when a value is provided
    - All `response` properties have a test for the default value
    - All `response` properties have a test for when a value is provided
    - All `response` functions have a test to check it is chainable, will close #58 if still stale
- Implement changes to resolve #72
- Implement changes to resolve #73 

Updated GitHub Action to publish as a beta, updated version number.

### CODE REVIEW

File: src/test/request/provided/provided.express.test.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  })
+
+  test('req.ips can be provided', () => {
+    const req = getMockReq({ ips: providedStringArray })
+
+    expect(req.ips).toBeDefined()
+    expect(req.ips).toBe(providedStringArray)
+  })
+
+  test('req.ips can be provided', () => {
```
Comment (bikk-uk) [inline src/test/request/provided/provided.express.test.ts]: Test name

File: src/test/request/provided/provided.express.test.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+  test('req.is can be provided', () => {
+    const req = getMockReq({ is: providedFunction })
+
+    expect(req.is).toBeDefined()
+    expect(req.is).toBe(providedFunction)
+  })
+
+  test('req.app can be provided', () => {
+    const req = getMockReq({ app: providedObject as unknown as Request['app'] })
```
Comment (bikk-uk) [inline src/test/request/provided/provided.express.test.ts]: Double check now `Partial`

File: src/test/request/provided/provided.http.IncomingMessage.test.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+  test('req.complete can be provided', () => {
+    const req = getMockReq({ complete: providedBoolean })
+
+    expect(req.complete).toBeDefined()
+    expect(req.complete).toBe(providedBoolean)
+  })
+
+  test('req.connection can be provided', () => {
+    const req = getMockReq({ connection: providedObject as unknown as Request['connection'] })
```
Comment (bikk-uk) [inline src/test/request/provided/provided.http.IncomingMessage.test.ts]: Double check now `Partial`

File: src/test/response/defaults/defaults.http.ServerResponse.test.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  })
+
+  test('res.writeHead is chainable', () => {
+    const { res } = getMockRes()
+
+    // it returns itself (is chainable)
+    expect(res.writeHead(200)).toBe(res)
+  })
+
+  test('res.writeProcessing is a mocked function and is not chainable', () => {
```
Comment (bikk-uk) [inline src/test/response/defaults/defaults.http.ServerResponse.test.ts]: Test name

File: src/test/response/defaults/defaults.http.ServerResponse.test.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    const { res } = getMockRes()
+
+    expect(res.writeProcessing).toBeDefined()
+    expect(typeof res.writeProcessing).toBe('function')
+    expect((res.writeProcessing as jest.Mock).getMockName()).toBe('writeProcessing mock default')
+  })
+
+  test('res.writeProcessing is a mocked function and is not chainable', () => {
```
Comment (bikk-uk) [inline src/test/response/defaults/defaults.http.ServerResponse.test.ts]: Test name

File: src/test/response/response.test.ts [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+  test('it contains the correct amount of properties', () => {
+    const mockRes = getMockRes()
+
+    // the response contains values for res and next and clear functions
+    expect(mockRes).toBeInstanceOf(Object)
+    expect(Object.keys(mockRes).length).toBe(4)
+  })
+
+  test('the res object is provided', () => {
+    const mockRes = getMockRes()
```
Comment (bikk-uk) [inline src/test/response/response.test.ts]: Consistency

**Review by bikk-uk — CHANGES_REQUESTED** [review]

File: src/test/request/provided/provided.express.test.ts (hunk 1 above)
Comment (aboyce) [reply src/test/request/provided/provided.express.test.ts]: Done

File: src/test/request/provided/provided.express.test.ts (hunk 2 above)
Comment (aboyce) [reply src/test/request/provided/provided.express.test.ts]: Done, used existing `providedApp`.

File: src/test/request/provided/provided.http.IncomingMessage.test.ts (hunk 3 above)
Comment (aboyce) [reply src/test/request/provided/provided.http.IncomingMessage.test.ts]: Done, used existing `providedSocket`.

File: src/test/response/defaults/defaults.http.ServerResponse.test.ts (hunk 4 above)
Comment (aboyce) [reply src/test/response/defaults/defaults.http.ServerResponse.test.ts]: Updated.

File: src/test/response/defaults/defaults.http.ServerResponse.test.ts (hunk 5 above)
Comment (aboyce) [reply src/test/response/defaults/defaults.http.ServerResponse.test.ts]: Updated.

File: src/test/response/response.test.ts (hunk 6 above)
Comment (aboyce) [reply src/test/response/response.test.ts]: Fixed.

**Review by bikk-uk — APPROVED** [review]

**Pull request merged**